### PR TITLE
feat: UpdateEmailUseCase

### PR DESCRIPTION
**Pull request by MohamadJaara**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

user need to update self email address

### Solutions

implement UpdateEmailUseCase and the needed Repository functions

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/UpdateEmailUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import com.wire.kalium.logic.feature.user.UpdateEmailUseCase.Result
+import com.wire.kalium.logic.functional.fold
+import com.wire.kalium.network.exceptions.KaliumException
+import com.wire.kalium.network.exceptions.isInvalidEmail
+import com.wire.kalium.network.exceptions.isKeyExists
+
+/**
+ * Updates the user's email address.
+ * keep in mind that this use case does not update the email immediately, it only sends a request to the backend.
+ * the use need to confirm the email change by clicking on the link in an email they receive.
```
Comment (saleniuk) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/user/UpdateEmailUseCase.kt:31]: typo: should be "the user" 😄

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/UserDAO.kt [hunk 2]
```diff
@@ -18,6 +18,7 @@
 
 package com.wire.kalium.persistence.dao
 
+import com.wire.kalium.persistence.dao.ManagedByEntity.WIRE
```
Comment (saleniuk) [inline persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/UserDAO.kt:21]: Is this import required? 🤔

**Review by saleniuk — APPROVED** [review]

File: persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/UserDAO.kt (hunk 2 above)
Comment (MohamadJaara) [reply persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/UserDAO.kt:21]: Not really, I have optimize imports enabled when doing a commit from AS, and for some reason it thought that this was an optimization

**Review by ohassine — APPROVED** [review]
LGTM ✨

**Pull request merged**